src/compiler/checker.ts

```diff
@@ -20645,10 +20645,20 @@ namespace ts {
                 }
                 return apparentType;
             }
-            const prop = getPropertyOfType(apparentType, right.escapedText);
+            let prop = getPropertyOfType(apparentType, right.escapedText);
             if (isIdentifier(left) && parentSymbol && !(prop && isConstEnumOrConstEnumOnlyModule(prop))) {
                 markAliasReferenced(parentSymbol, node);
             }
+
+            // Check for properties added via mixer-style base types also
+            if (!prop && typeIsInterfaceType(apparentType)) {
+                for (const base of apparentType.resolvedBaseTypes) {

```

Comment:
Properties are supposed to get merged down into types that inherit from them - check out `resolveObjectTypeMembers`. You should _never_ need to traverse "base types" to find a property like this, as base types are a syntactic construct only (we're a structural typesystem, after all). I would check out why the property isn;t being included there, instead.

Reply:
I took a look at this, and the issue is that there’s a circularity resolving the base types `IQuark` while checking its declaration. As soon as we start resolving base types, we set `type.resolvedBaseTypes` to an empty array, which prevents actually _crashing_ due to this circularity. During that resolution, something needs the resolved members of `IQuark`, which sees no base types because they’re still resolving, so it has its `members` set to an empty array.

Later, when we check `a.value` against `IQuark`, its base types _have_ been resolved correctly, but because we’ve cached an empty `members` map, we don’t try to re-resolve them, so they appear empty.

The circularity that occurred was really deep (happened during the instantiation of the conditional `InstanceType`) and I didn’t put tons of effort into understanding exactly why it happened. It does seem like you should be able to substitute `A` in `interface A extends B<T> {}` for `B<T>`, so it seems like the best way forward is to see if the circularity can be eliminated. I _was_ able to fix the error by clearing `type.members` if it changed while resolving base types since that seems like a good indicator that something went awry, but that feels pretty hacky.

Reply:
Oh, hang on—`observe()` returns an `IQuark`, so the circularity is real and obvious! It’s even called out explicitly in https://github.com/microsoft/TypeScript/issues/29872:

> The workaround exists - to use different form of creating the type for the standalone mixin class - with interfaces... But this notation seems to drive crazy the IDE (the language server under the hood?). It stop finding usages of properties, show many false type errors etc.

Yep, because there’s a circularity that we falsely resolve and don’t report. I think the bug is that we _don’t_ report a circularity error— the “workaround” is the bug.

And my earlier comment:

> you should be able to substitute `A` in `interface A extends B<T> {}` for `B<T>`

It turns out the simple form `B<T>` (`Mixin<typeof CQuark>` in this test case) simply resolves to `any` in this case without triggering any errors, so that doesn’t actually work either, so I feel better about this interface example being an error. (I guess in the type reference case, we detect the circularity but fail to emit a noImplicitAny error on the way back out.)

Reply:
I think we should report circularity errors when resolving base types, but I would push this to 3.8 to avoid a last-minute breaking change.

Reply:
> I think we should report circularity errors when resolving base types

I agree

Reply:
To clarify, I think it may be possible to fix #29872 in the future such that this scenario wouldn’t trigger the circularity error, but as it stands now, it’s silently broken, which I think is worse.

Reply:
@andrewbranch the fix for #29872 is very anticipated, if only would you know how much effort our team has put into silencing this kind of errors, which seems to be perfectly valid for regular classes. I think mixin pattern is already established enough to be supported first-class. Looking forward.

Reply:
@samuraijack The fix you want actually almost happened in #33050, but was removed because the performance cost was too high:

> The specific change made by this PR is that type arguments are permitted to make circular references in aliased types of the following kinds:
>
> Instantiations of generic class and interface types (for example `Array<Foo>`).
> Array types (for example `Foo[]`).
> Tuple types (for example `[string, Foo?]`).

then [later](https://github.com/microsoft/TypeScript/pull/33050#issuecomment-525471967)...

> Performance tests above show a worst case 5% check time increase and 10% memory consumption increase. That's a bit too expensive. With the latest commit we only defer type argument resolution for aliased array and tuple types. That should improve the numbers.

This is actually observable in your example in 3.7 by [changing the `SampleMixin1` and `SampleMixin2` to be tuple or array types](https://www.typescriptlang.org/play/#code/LAKApgHgDg9gTgFwAQGMYDsDOyDKBDAWygBswBZASwgvQEYkBeJAHgBUlIEx0ATTJAILoAngGEM2OAFcUCeMxgAjAFZhZAPnUAKRXkxgAXElYBKRutApie-viKlK1Ohwhde-XfqQBvUEn9IeOgwCAAWYHAAhEZ2JORUNABMoAC+oJCwiEgIwlBgSLEOCc5MANqONMw5eTAAZgWEcRV06gC6oOnQ8MhoWLiNRU6JjCzsnNx8giLifdKy8kqqGtqe+Uam5pbWmLYD8UMubpOrPn4BQSHhUTF7zbSpnZnI1fmF+0kj5cVVuWB1DfZ3uhEm0OiBQAB6CFITBSKBPGgAcyQPDU1jgeAQFAkSFq8CQBGKSCgmK4cHQj262V+U2EADEpOhZNj0MwBCMgsJ1AEAkwtAA6QU0KBSBBGTmlVpmBjcgSUrIvWkzSQyORwNkckTc-xMdBgADuSAFQvQIrFgREkulsvlzxpzTYhwm-CE9MZzIw3KYAEk+kEUGBWL9mAAlMAIKTkoN5NiaIA).

I know this is one of these so-close-yet-so-far things—I’m not suggesting you can get by with making all your types tuples. I’m pointing this out because it will help you ask for the right thing. This has nothing to do with mixins; this is all about circular type references. You want circular type references to work for instantiations of generic types.

Reply:
Oh, #33050 looks like exactly what I need. And its labeled as "Merged" with Milestone 3.7.0? So its already there?

Reply:
#33050 _originally_ fixed your case but it was scoped down due to performance reasons. You can see what I mean in the [playground link](https://www.typescriptlang.org/play/#code/LAKApgHgDg9gTgFwAQGMYDsDOyDKBDAWygBswBZASwgvQEYkBeJAHgBUlIEx0ATTJAILoAngGEM2OAFcUCeMxgAjAFZhZAPnUAKRXkxgAXElYBKRutApie-viKlK1Ohwhde-XfqQBvUEn9IeOgwCAAWYHAAhEZ2JORUNABMoAC+oJCwiEgIwlBgSLEOCc5MANqONMw5eTAAZgWEcRV06gC6oOnQ8MhoWLiNRU6JjCzsnNx8giLifdKy8kqqGtqe+Uam5pbWmLYD8UMubpOrPn4BQSHhUTF7zbSpnZnI1fmF+0kj5cVVuWB1DfZ3uhEm0OiBQAB6CFITBSKBPGgAcyQPDU1jgeAQFAkSFq8CQBGKSCgmK4cHQj262V+U2EADEpOhZNj0MwBCMgsJ1AEAkwtAA6QU0KBSBBGTmlVpmBjcgSUrIvWkzSQyORwNkckTc-xMdBgADuSAFQvQIrFgREkulsvlzxpzTYhwm-CE9MZzIw3KYAEk+kEUGBWL9mAAlMAIKTkoN5NiaIA) from my comment above.

Reply:
I see.

Yup, how is the trick with tuples different from using the `interface`? I tried to continue your example, but indeed stuck when trying to access the nested property: [link](https://www.typescriptlang.org/play/#code/LAKApgHgDg9gTgFwAQGMYDsDOyDKBDAWygBswBZASwgvQEYkBeJAHgBUlIEx0ATTJAILoAngGEM2OAFcUCeMxgAjAFZhZAPnUAKRXkxgAXElYBKRutApie-viKlK1Ohwhde-XfqQBvUEn9IeOgwCAAWYHAAhEZ2JORUNABMoAC+oJCwiEgIwlBgSLEOCc5MANqONMw5eTAAZgWEcRV06gC6oOnQ8MhoWLiNRU6JjCzsnNx8giLifdKy8kqqGtqe+Uam5pbWmLYD8UMubpOrPn4BQSHhUTF7zbSpnZnI1fmF+0kj5cVVuWB1DfZ3uhEm0OiBethsmBIUwtHgboC7qUAAytMwMdQ+JBnfx4AB0FzCEQJwSJcDBAUCJMuERRrWpZIeIFAAHoWUhMFIoE8aABzJA8NTWOB4BAUCRIWrwJAEYpIKCirhwdCPbrZX5TYQAMSk6Fk4vQzAEIyCwkxlJGWjx1poUCkCCMptKaPMglVWRempmkhkcjgRpNInNI3QYAA7kgrTb0HaHYERM70ZiBO7nhrmmxDhN+EJtbr9RhMUwAJJ9IIoMCsX7MABKYAQUmVVbybE0QA)

Perhaps its possible to add some additional type, that will resolve the tuple type.. Something [like this](https://www.typescriptlang.org/play/#code/LAKApgHgDg9gTgFwAQGMYDsDOyDKBDAWygBswBZASwgvQEYkBeJAHgBUlIEx0ATTJAILoAngGEM2OAFcUCeMxgAjAFZhZAPnUAKRXkxgAXElYBKRutApie-viKlK1Ohwhde-XfqQBvUEn9IeOgwCAAWYHAAhEZ2JORUNABMiaAAvqCQsIhICMJQYEixDgnOTADajjTMufkwAGaFhHGVdOoAuqAZ0PDINQVF8U609EwDLbRlAAwdIF1ZyGhYuE3FTomMLOyc3HyCIuJL0rLySqoa2p4FRqbmltaYtiuDSS5uu5c+fgFBIeFRMU9xrQ0nMejk8v1ASV1uUWtUIfVGvZnuhEu1OuButk+kjmtCYbjVkkpjNLBJemBsBstHgAcjxiSzAx1D4kF9-HgAHQ-MIRbnBXlwDEBQL834REliwUg2YgAD0cqQmCkUHmNAA5kgeGprHA8AgKBIkHV4EgCCUkFB9Vw4OhQdiIXthAAxKToWSG9DMAQbILCFki6mc4M0KBSBBGP1lNpMlkCe29R1CMTko5yODe30iAMbdBgADuSC0wc5ofDkZE0djggT4PySDhW1cO34ydd7oNGBZTAAkksgigwKwIcwAEpgBBSW3D-JsTRAA) - which currently fails.

Reply:
It’s not possible. The trick with tuples/arrays is unique to tuples/arrays because they’re _explicitly_ allowed to be recursive via #33050. If you pull the element out of the tuple/array, it’s no longer a tuple/array; it’s an indexed access type and it no longer falls into the special case set up by #33050. As far as I can tell, the only way to get what you’re really asking for is to expand the cases where recursive type aliases are allowed to include generic instantiations.

Reply:
> As far as I can tell, the only way to get what you’re really asking for is to expand the cases where recursive type aliases are allowed to include generic instantiations.

@andrewbranch Is this something that #33050 provides? If so, can it be enabled by some flag? #33050 says its not enabled globally, because of the 5% performance hit. I believe for many people, the value of the circular types is much more than those 5%.

Reply:
> Is this something that #33050 provides?

No, it is not.